src/libpgmoneta/utils.c

```diff
@@ -3003,6 +3004,198 @@ pgmoneta_is_file_archive(char* file_path)
    return false;
 }
 
+int
+pgmoneta_token_bucket_init(struct token_bucket* tb, unsigned long burst, unsigned long max_rate, int every)
+{
+   if (tb != NULL)
+   {
+      tb->burst = burst;
+      atomic_init(&tb->cur_tokens, burst);
+      tb->max_rate = max_rate;
+      tb->every = every;
+      atomic_init(&tb->last_time, (unsigned long)time(NULL));
+      return 0;
+   }
+   return 1;
+}
+
+void
+pgmoneta_token_bucket_destroy(struct token_bucket* tb)
+{
+   if (tb)
+   {
+      free(tb);
+   }
+}
+
+int
+pgmoneta_token_bucket_add(struct token_bucket* tb)
+{
+   unsigned long diff;
+   unsigned long expected_tokens;
+   unsigned long new_tokens;
+   unsigned long expected_time;
+   unsigned long cur_time;
+
+   diff = (unsigned long) time(NULL) - atomic_load(&tb->last_time);
+
+   if (diff <= 0)
```

Comment:
Then probably this should be diff <= tb->every

Reply:
Yes, you are right. I always use 1 second for testing and forgot about that.